common/persistence/elasticsearchVisibility.go

```diff
@@ -0,0 +1,395 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package persistence
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/olivere/elastic"
+	"github.com/pkg/errors"
+	"github.com/uber-common/bark"
+	workflow "github.com/uber/cadence/.gen/go/shared"
+	"github.com/uber/cadence/common"
+	es "github.com/uber/cadence/common/elasticsearch"
+	"github.com/uber/cadence/common/logging"
+)
+
+const esPersistenceName = "elasticsearch"
+
+type (
+	esVisibilityManager struct {
+		esClient es.Client
+		index    string
+		logger   bark.Logger
+	}
+
+	esVisibilityPageToken struct {
+		From int
+	}
+
+	visibilityRecord struct {
+		WorkflowID    string
+		RunID         string
+		WorkflowType  string
+		StartTime     int64
+		CloseTime     int64
+		CloseStatus   workflow.WorkflowExecutionCloseStatus
+		HistoryLength int64
+	}
+)
+
+var _ VisibilityManager = (*esVisibilityManager)(nil)
+
+var (
+	errOperationNotSupported = errors.New("operation not support")
+)
+
+// NewElasticSearchVisibilityManager create a visibility manager connecting to ElasticSearch
+func NewElasticSearchVisibilityManager(esClient es.Client, index string, logger bark.Logger) VisibilityManager {
+	return &esVisibilityManager{
+		esClient: esClient,
+		index:    index,
+		logger:   logger.WithField(logging.TagWorkflowComponent, logging.TagValueESVisibilityManager),
+	}
+}
+
+func (v *esVisibilityManager) Close() {}
+
+func (v *esVisibilityManager) GetName() string {
+	return esPersistenceName
+}
+
+func (v *esVisibilityManager) RecordWorkflowExecutionStarted(request *RecordWorkflowExecutionStartedRequest) error {
+	return errOperationNotSupported
```

Comment:
Can we have Kafka publisher code moved here.

Reply:
ok, will do in separate PR.